pkcs11/tests/ecdh_sp800_test.c

```diff
@@ -0,0 +1,509 @@
+CK_BYTE P256_PARAMS[] = {0x06, 0x08, 0x2a, 0x86, 0x48,
+                         0xce, 0x3d, 0x03, 0x01, 0x07};
+CK_BYTE P384_PARAMS[] = {0x06, 0x05, 0x2b, 0x81, 0x04, 0x00, 0x22};
+CK_BYTE P521_PARAMS[] = {0x06, 0x05, 0x2b, 0x81, 0x04, 0x00, 0x23};
+
+static CK_FUNCTION_LIST_PTR p11;
+static CK_SESSION_HANDLE session;
+
+char *CURVES[] = {"secp224r1", "prime256v1", "secp384r1", "secp521r1"};
+CK_BYTE *CURVE_PARAMS[] = {P224_PARAMS, P256_PARAMS, P384_PARAMS, P521_PARAMS};
+CK_ULONG CURVE_LENS[] = {sizeof(P224_PARAMS), sizeof(P256_PARAMS),
+                         sizeof(P384_PARAMS), sizeof(P521_PARAMS)};
+int CURVE_COUNT = sizeof(CURVE_PARAMS) / sizeof(CURVE_PARAMS[0]);
+CK_ULONG KDFS[] = {CKD_NULL, CKD_YUBICO_SHA1_KDF_SP800,
+                   CKD_YUBICO_SHA256_KDF_SP800, CKD_YUBICO_SHA384_KDF_SP800,
+                   CKD_YUBICO_SHA512_KDF_SP800};
+int KDFS_LEN = sizeof(KDFS) / sizeof(CK_ULONG);
+
+static void success(const char *message) { printf("%s. OK\n", message); }
+
+static void fail(const char *message) { printf("%s. FAIL!\n", message); }
+
+static void generate_keypair_yh(CK_BYTE *curve, CK_ULONG curve_len,
+                                CK_OBJECT_HANDLE_PTR publicKeyPtr,
+                                CK_OBJECT_HANDLE_PTR privateKeyPtr) {
+  CK_MECHANISM mechanism = {CKM_EC_KEY_PAIR_GEN, NULL_PTR, 0};
+
+  CK_BBOOL ck_true = CK_TRUE;
+
+  CK_OBJECT_CLASS pubkey_class = CKO_PUBLIC_KEY;
+  CK_OBJECT_CLASS privkey_class = CKO_PRIVATE_KEY;
+  CK_KEY_TYPE key_type = CKK_EC;
+  char *label = "ecdhtest";
+
+  CK_ATTRIBUTE publicKeyTemplate[] = {{CKA_CLASS, &pubkey_class,
+                                       sizeof(pubkey_class)},
+                                      {CKA_VERIFY, &ck_true, sizeof(ck_true)},
+                                      {CKA_KEY_TYPE, &key_type,
+                                       sizeof(key_type)},
+                                      {CKA_LABEL, label, strlen(label)},
+                                      {CKA_EC_PARAMS, curve, curve_len}};
+
+  CK_ATTRIBUTE privateKeyTemplate[] = {{CKA_CLASS, &privkey_class,
+                                        sizeof(privkey_class)},
+                                       {CKA_LABEL, label, strlen(label)},
+                                       {CKA_DERIVE, &ck_true, sizeof(ck_true)}};
+
+  if ((p11->C_GenerateKeyPair(session, &mechanism, publicKeyTemplate, 5,
+                              privateKeyTemplate, 3, publicKeyPtr,
+                              privateKeyPtr)) != CKR_OK) {
+    fail("Failed to generate EC key pair on YubiHSM");
+    exit(EXIT_FAILURE);
+  }
+  success("Generated EC key pair on YubiHSM");
+}
+
+static EVP_PKEY *generate_keypair_openssl(const char *curve) {
+  EVP_PKEY *pkey = NULL;
+  EC_KEY *eckey = NULL;
+  OpenSSL_add_all_algorithms();
+  int eccgrp = OBJ_txt2nid(curve);
+  eckey = EC_KEY_new_by_curve_name(eccgrp);
+  if (!(EC_KEY_generate_key(eckey))) {
+    fail("Failed to generate EC keypair with openssl");
+  } else {
```

Comment:
Dont use an else here, it's enought that it errors out

Reply:
Fixed